File: org.boris.xlloop/servers/ruby/server-example.rb
require "xlloop"

class TestHandler
	def initialize
		@arrr = [('a'..'z'),('A'..'Z')].map{|i| i.to_a}.flatten
		@functionInfo = createFunctionInfo
	end
	
	def invoke(context, name, args)
	  print "Invoking #{name} ", args.inspect, "\n"
		case name
		when "RandTest" then # Test encoding of different types and different lengths
			a = Array.new
			(1 + rand(50)).times do 
				case rand(4)
				when 0 then a.push((0..50).map{ @arrr[rand(@arrr.length)]  }.join)
				when 1 then a.push(4000 * rand())
				when 2 then a.push(rand(2)==1 ? true : false)
				when 3 then a.push(rand(435345))
				end
			end
			return a
		when "ArrayTest" then # Test encoding of arrays
			a = Array.new
			for i in 1..15
				b = Array.new
				for j in 1..rand(10)
					b.push(j * i)
				end
				a.push(b)
			end
			return a
		when "ArgsTest" then
			return args[0]
    when "GetEmptyString" then
      return ""
		when "org.boris.xlloop.GetFunctions" then # Registers functions in excel 
			return @functionInfo
		else
			"#Unknown function"
		end
	end
	
	def createFunctionInfo
		a = Array.new
		a.push(FunctionInformation.new("RandTest", 
			"Test encoding of different types and different lengths"))
		a.push(FunctionInformation.new("ArrayTest", "Test encoding of arrays"))
		f3 = FunctionInformation.new("ArgsTest", "Testing inputting args")
		f3.addArg("anything", "Test this one")
		f3.addArg("else", "Whatever you like")
		f3.category = "Testing"
		a.push(f3)
		return a
	end
end

f = XLLoopServer.new(TestHandler.new, 5470)
puts "XLLoop Ruby Server Example..."
f.start

File: org.boris.xlloop/servers/python/server-example.py
from xlloop import *
import string, random, array

class TestHandler:
    def invoke(self, context, name, args):
        print(name)
        if name == 'ArgsTest':
            if(len(args) > 0):
                print(args[0])
                return args[0]
            else:
                return XLError(XL_ERROR_NA)
        
        elif name == 'RandTest':
            rlen = int(random.random() * 50 + 2)
            a = []
            for i in xrange(rlen):
                v = int(random.random() * 4)
                if v == 0:
                    a.append("".join(random.sample(string.letters+string.digits, 50)))
                elif v == 1:
                    a.append(int(random.random() * 1000))
                elif v == 2:
                    if int(random.random() * 2) == 1:
                        a.append(True)
                    else:
                        a.append(False)
                else:
                    a.append(random.random() * 1000)
            return a
        elif name == 'GetEmptyString':
            return ''
        else:
            return [random.random() * 1000, 'Hello World!', 23.3, 11111]
        
h = TestHandler()
xs = XLLoopServer(h, 5460)
xs.start()

